<a href="https://colab.research.google.com/github/iam-fern/DPDM/blob/master/Association_Rule.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
# connect with google drive
import pandas as pd
from google.colab import drive
drive.mount('/content/drive')
import os
base_path = '/content/drive/My Drive/DWDM_material'
os.listdir(base_path)

Mounted at /content/drive


['test.csv',
 'metadata.xlsx',
 'search_click.csv',
 'w_restaurant.csv',
 'Copy of search_click.csv',
 '03 สังคม.xls',
 '52835793_259306464964511_3851804852342489088_n.png',
 'reduced_marketbasket.csv',
 'Copy of w_restaurant.csv',
 'Preprocessing02.html',
 'teach02.csv',
 'teach.csv',
 'baboon.jpg',
 'data2.csv',
 'data1.gsheet',
 'data3.gsheet',
 'data3.xlsx',
 'data3.csv',
 'data1.xlsx',
 'data1.csv',
 'Local.xlsx']

In [2]:
data = pd.read_csv(os.path.join(base_path,'reduced_marketbasket.csv'),encoding = 'unicode_escape')      # มีอักษร latin(ลาติน) ในไฟล์เลยต้องเพิ่ม encoding latin1
data

InvoiceNo StockCode  ... CustomerID         Country
0         536365    85123A  ...    17850.0  United Kingdom
1         536365     71053  ...    17850.0  United Kingdom
2         536365    84406B  ...    17850.0  United Kingdom
3         536365    84029G  ...    17850.0  United Kingdom
4         536365    84029E  ...    17850.0  United Kingdom
...          ...       ...  ...        ...             ...
541904    581587     22613  ...    12680.0          France
541905    581587     22899  ...    12680.0          France
541906    581587     23254  ...    12680.0          France
541907    581587     23255  ...    12680.0          France
541908    581587     22138  ...    12680.0          France

[541909 rows x 8 columns]

### Ex. ให้หาว่าแต่ละประเทศ ขายสินค้ารวมกันทั้งหมดได้ประเทศละกี่ชิ้น

In [3]:
data.groupby('Country').sum().sort_values('Quantity')

Quantity     UnitPrice    CustomerID
Country                                                   
Saudi Arabia                75  2.411000e+01  1.256500e+05
Bahrain                    260  8.657000e+01  2.100270e+05
RSA                        352  2.481000e+02  7.218680e+05
Brazil                     356  1.426000e+02  4.086080e+05
Lebanon                    386  2.424400e+02  5.743800e+05
European Community         497  2.940500e+02  9.215880e+05
Czech Republic             592  8.815000e+01  3.834300e+05
Lithuania                  652  9.944000e+01  5.366200e+05
Malta                      944  6.660100e+02  2.158496e+06
United Arab Emirates       982  2.298900e+02  1.018952e+06
USA                       1034  6.449800e+02  3.672086e+06
Greece                    1556  7.132900e+02  2.008584e+06
Iceland                   2458  4.812100e+02  2.247154e+06
Canada                    2763  9.105800e+02  2.615483e+06
Unspecified               3300  1.204010e+03  3.348046e+06
Poland                    3653  1.422270e+03  4.341972e+06
Israel                    4353  1.079040e+03  3.164467e+06
Hong Kong                 4769  1.224150e+04  0.000000e+00
Austria                   4827  1.701520e+03  5.021102e+06
Singapore                 5234  2.510889e+04  2.918376e+06
Cyprus                    6317  3.920070e+03  7.715880e+06
Italy                     7999  3.879390e+03  1.015666e+07
Denmark                   8188  1.266950e+03  4.876734e+06
Channel Islands           9479  3.738550e+03  1.128522e+07
Finland                  10666  3.786850e+03  8.699324e+06
Portugal                 16180  1.303754e+04  1.886480e+07
Norway                   19247  6.529060e+03  1.350765e+07
Belgium                  23152  7.540130e+03  2.571829e+07
Japan                    25218  8.148600e+02  4.567292e+06
Spain                    26824  1.263345e+04  3.268929e+07
Switzerland              30325  6.813690e+03  2.377592e+07
Sweden                   35637  1.806830e+03  6.790083e+06
Australia                83653  4.054750e+03  1.569300e+07
France                  110480  4.303199e+04  1.076489e+08
Germany                 117448  3.766600e+04  1.200751e+08
EIRE                    142637  4.844719e+04  1.103917e+08
Netherlands             200128  6.492550e+03  3.419054e+07
United Kingdom         4263829  2.245715e+06  5.626433e+09

In [4]:
data.groupby('InvoiceNo').sum()

Quantity  UnitPrice  CustomerID
InvoiceNo                                 
536365           40      27.37    124950.0
536366           12       3.70     35700.0
536367           83      58.24    156564.0
536368           15      19.10     52188.0
536369            3       5.95     13047.0
...             ...        ...         ...
C581484      -80995       2.08     16446.0
C581490         -23       2.78     28794.0
C581499          -1     224.69     15498.0
C581568          -5      10.95     15311.0
C581569          -6       2.50     34630.0

[25900 rows x 3 columns]

In [5]:
transacs = data.groupby('InvoiceNo')['StockCode'].apply(list)     #เอามาใส่ใน list โดยไม่ sum หรือ count
transacs

InvoiceNo
536365     [85123A, 71053, 84406B, 84029G, 84029E, 22752,...
536366                                        [22633, 22632]
536367     [84879, 22745, 22748, 22749, 22310, 84969, 226...
536368                          [22960, 22913, 22912, 22914]
536369                                               [21756]
                                 ...                        
C581484                                              [23843]
C581490                                       [22178, 23144]
C581499                                                  [M]
C581568                                              [21258]
C581569                                       [84978, 20979]
Name: StockCode, Length: 25900, dtype: object

## Apriori

http://zaxrosenberg.com/unofficial-apyori-documentation/

In [6]:
!pip install apyori     # แพ็คเก็จที่คนไม่ค่อยใช้ต้องทำแบบนี้ก่อนทุกครั้งก่อนใช้งาน  #ติดตั้งไลบราลีก่อน

  Created wheel for apyori: filename=apyori-1.1.2-cp36-none-any.whl size=5975 sha256=0560477eb40ebb7878152c1ad6afb3c6ba5484c8bd81212fc7485fd6f19ded34
  Stored in directory: /root/.cache/pip/wheels/5d/92/bb/474bbadbc8c0062b9eb168f69982a0443263f8ab1711a8cad0
Successfully built apyori


In [7]:
from apyori import apriori

In [8]:
out = list(apriori(transacs, min_support = 0.03, min_confidence = 0.1))     #ครอบ list เพราะมันไม่ได้ออกมาแบบ list มันอ่านยาก
out                                                                         #RelationRecord 
                                                                            #items ก็คือ คู่อันดับที่เราสนใจ
                                                                            #support ก็คือ ความน่าจะเป็นที่เกิดคู่อันดับที่เราสนใจ
                                                                            #ordered_statistics ซื้ออันแรกแล้วซื้ออันหลัง มี confident คือเปอร์เซ็นตามที่แสดง 
                                                                            #lift ดูความสัมพันธ์เมื่อกำหนดเงื่อนไข

[RelationRecord(items=frozenset({'22386', '85099B'}), support=0.03216216216216216, ordered_statistics=[OrderedStatistic(items_base=frozenset({'22386'}), items_add=frozenset({'85099B'}), confidence=0.6766856214459789, lift=8.20897311262335), OrderedStatistic(items_base=frozenset({'85099B'}), items_add=frozenset({'22386'}), confidence=0.39016393442622943, lift=8.20897311262335)]),
 RelationRecord(items=frozenset({'22699', '22697'}), support=0.03027027027027027, ordered_statistics=[OrderedStatistic(items_base=frozenset({'22697'}), items_add=frozenset({'22699'}), confidence=0.7417218543046358, lift=17.1523178807947), OrderedStatistic(items_base=frozenset({'22699'}), items_add=frozenset({'22697'}), confidence=0.7, lift=17.1523178807947)])]

In [ ]:
out[0]

RelationRecord(items=frozenset({'22386', '85099B'}), support=0.03216216216216216, ordered_statistics=[OrderedStatistic(items_base=frozenset({'22386'}), items_add=frozenset({'85099B'}), confidence=0.6766856214459789, lift=8.20897311262335), OrderedStatistic(items_base=frozenset({'85099B'}), items_add=frozenset({'22386'}), confidence=0.39016393442622943, lift=8.20897311262335)])

## Quiz หา k-itemset ที่มีความน่าสนใจ (โดยพิจารณาลูกค้าเป็นรายคน) พร้อมทั้งอธิบายว่าน่าสนใจยังไง

In [9]:
transacs = data.groupby('CustomerID')['StockCode'].apply(list)       #เอามาใส่ใน list โดยไม่ sum หรือ count
transacs

CustomerID
12346.0                                       [23166, 23166]
12347.0    [85116, 22375, 71477, 22492, 22771, 22772, 227...
12348.0    [84992, 22951, 84991, 84991, 21213, 21213, 226...
12349.0    [23112, 23460, 21564, 21411, 21563, 22131, 221...
12350.0    [21908, 22412, 79066K, 79191C, 22348, 84086C, ...
                                 ...                        
18280.0    [82484, 22180, 22467, 22725, 22727, 22495, 223...
18281.0    [22037, 22716, 22028, 23007, 23008, 23209, 22467]
18282.0    [21270, 23187, 23295, 22089, 21108, 21109, 224...
18283.0    [22356, 20726, 22384, 22386, 20717, 20718, 850...
18287.0    [22755, 22754, 22753, 22756, 22758, 22757, 227...
Name: StockCode, Length: 4372, dtype: object

In [ ]:
out = list(apriori(transacs, min_support = 0.1, min_confidence = 0.2))
out

[RelationRecord(items=frozenset({'22423'}), support=0.20288197621225984, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22423'}), confidence=0.20288197621225984, lift=1.0)])]

### อาจารย์ทำ

In [ ]:
out = list(apriori(transacs, min_support = 0.07, min_confidence = 0.7))
out

[RelationRecord(items=frozenset({'85123A', '21733'}), support=0.07136322049405307, ordered_statistics=[OrderedStatistic(items_base=frozenset({'21733'}), items_add=frozenset({'85123A'}), confidence=0.8103896103896104, lift=4.129397874852421)]),
 RelationRecord(items=frozenset({'22699', '22423'}), support=0.07410795974382434, ordered_statistics=[OrderedStatistic(items_base=frozenset({'22699'}), items_add=frozenset({'22423'}), confidence=0.7587822014051523, lift=3.740017795426523)]),
 RelationRecord(items=frozenset({'22699', '22697'}), support=0.07410795974382434, ordered_statistics=[OrderedStatistic(items_base=frozenset({'22697'}), items_add=frozenset({'22699'}), confidence=0.8307692307692307, lift=8.506143037290578), OrderedStatistic(items_base=frozenset({'22699'}), items_add=frozenset({'22697'}), confidence=0.7587822014051523, lift=8.506143037290578)])]

คนจะซื้อ {'85123A'} และ {'21733'} พร้อมกัน คิดเป็น 7.136322049405307 % และ คนที่ซื้อ {'85123A'} จะซื้อ {'21733'} คิดเป็น 81.03896103896104 %

คนจะซื้อ {'22699'} และ {'22423'} พร้อมกัน คิดเป็น 7.410795974382434 % และ คนที่ซื้อ {'22699'} จะซื้อ {'22423'} คิดเป็น 75.87822014051523 %

คนจะซื้อ {'22699'} และ {'22697'} พร้อมกัน คิดเป็น 7.410795974382434 % และ คนที่ซื้อ {'22699'} จะซื้อ {'22697'} คิดเป็น 83.07692307692307 %
******************************************************************
เนื่องจากคนที่จะซื้อสินค้า {'21733', '85123A'} มีความน่าจะเป็นคือ 0.07136322049405307

เนื่องจากคนที่จะซื้อสินค้า {'22423', '22699'} มีความน่าจะเป็นคือ 0.07410795974382434

เนื่องจากคนที่จะซื้อสินค้า {'22697', '22699'} มีความน่าจะเป็นคือ 0.07410795974382434

และ lift > 1 ทุกคู่ ดังนั้นจะมีโปรโมชั่นให้ ถ้าลูกค้าซื้อสินค้าคู่ดังกล่าวสะสมครบ 10 ครั้ง ทางร้านจะแถมสินค้า 1 คู่ให้ โดยลูกค้าสามารถเลือกคู่ใดคู่หนึ่งได้ฟรี

In [ ]:
transacs = data.groupby('CustomerID')['Description'].apply(list)
transacs

CustomerID
12346.0    [MEDIUM CERAMIC TOP STORAGE JAR, MEDIUM CERAMI...
12347.0    [BLACK CANDELABRA T-LIGHT HOLDER, AIRLINE BAG ...
12348.0    [72 SWEETHEART FAIRY CAKE CASES, 60 CAKE CASES...
12349.0    [PARISIENNE CURIO CABINET, SWEETHEART WALL TID...
12350.0    [CHOCOLATE THIS WAY METAL SIGN, METAL SIGN NEI...
                                 ...                        
18280.0    [WOOD BLACK BOARD ANT WHITE FINISH, RETROSPOT ...
18281.0    [ROBOT BIRTHDAY CARD, CARD CIRCUS PARADE, PENN...
18282.0    [ANTIQUE CREAM CUTLERY CUPBOARD, FRENCH STYLE ...
18283.0    [CHARLOTTE BAG PINK POLKADOT, LUNCH BAG WOODLA...
18287.0    [SMALL PURPLE BABUSHKA NOTEBOOK , SMALL RED BA...
Name: Description, Length: 4372, dtype: object

In [ ]:
out = list(apriori(transacs, min_support = 0.07, min_confidence = 0.7))
out

[RelationRecord(items=frozenset({'GREEN REGENCY TEACUP AND SAUCER', 'ROSES REGENCY TEACUP AND SAUCER '}), support=0.07410795974382434, ordered_statistics=[OrderedStatistic(items_base=frozenset({'GREEN REGENCY TEACUP AND SAUCER'}), items_add=frozenset({'ROSES REGENCY TEACUP AND SAUCER '}), confidence=0.8307692307692307, lift=8.506143037290578), OrderedStatistic(items_base=frozenset({'ROSES REGENCY TEACUP AND SAUCER '}), items_add=frozenset({'GREEN REGENCY TEACUP AND SAUCER'}), confidence=0.7587822014051523, lift=8.506143037290578)]),
 RelationRecord(items=frozenset({'RED HANGING HEART T-LIGHT HOLDER', 'WHITE HANGING HEART T-LIGHT HOLDER'}), support=0.07136322049405307, ordered_statistics=[OrderedStatistic(items_base=frozenset({'RED HANGING HEART T-LIGHT HOLDER'}), items_add=frozenset({'WHITE HANGING HEART T-LIGHT HOLDER'}), confidence=0.8103896103896104, lift=4.129397874852421)]),
 RelationRecord(items=frozenset({'ROSES REGENCY TEACUP AND SAUCER ', 'REGENCY CAKESTAND 3 TIER'}), support=